cEP-0016.md

```diff
@@ -0,0 +1,455 @@
+shortlog_regex = ([^:]*|[^:]+: [A-Z0-9*].*)
+body_close_issue = True
+body_close_issue_full_url = True
+body_close_issue_on_last_line = True
+body_enforce_issue_reference = False
+```
+
+Project- [coala/coala-atom](https://github.com/coala/coala-atom)
+
+```
+[commit]
+bears = GitCommitBear
+```
+
+## Proposed Change
+
+This project would allow organizations to customize the `.coafile`
+to allow following settings in the `[commit]` section.
+
+```
+[commit]
+bears = GitCommitBear, CISkipInspectBear, GitMergeInspectBear, GitRevertInspectBear
+allow_skip_ci_build = False
+allow_git_merge_commit = False
+allow_git_revert_commit = True
+```
+
+GitHub Pull Request temporary merge commits would be inspected by making
+changes in the `GitCommitBear`. For other special commits, a meta bear
+called `VCSCommitBear` would be used to provide information about the HEAD
+commit like commit message, commit SHA, type of commit and list of modified,
+added and delete files as `HiddenResult` which will be used be other bears
+for inspection of each special commit separately.
+
+Here is the detailed implementation stepwise:
+
+1. We start by identifying different types of special commits namely GitHub Pull
+   Request temporary merge commits, `git revert` , `git merge` and commits that
+   disable CI builds.
+2. Each type of commit is then handled as a separate case.
+3. For GitHub Pull Request temporary merge commits:
+   - As a solution to this problem, settings will be added in `GitCommitBear`
+     that would identify GitHub Pull Request temporary merge commits and ignore
+     them.
+   - `GitCommitBear` would identify and check contents of the unmerged parent of
+     these merge commits.
+4. A meta bear called `VCSCommitBear` would be created that would return a
+   `HiddenResult` comprising of a tuple that will contain information about
+   the type of git commit, commit sha and lists of modified files, added and
+   deleted files.
+   This meta bear would be used to inspect each type of special commit.
+5. For commits that disable CI builds:
+   This would involve making a new bear named `CISkipInspectBear` to inspect
+   commits that disable CI builds. This bear would contain metadata about
+   various CI providers that are supported by coala such as Circle CI,
+   Travis CI, Semaphore CI etc. Most of the CI providers use sequences like
+   `[ci skip]` and `[skip ci]` for skipping CI build on commits. CI build
+   don't occur for commits that have either of these sequences anywhere in
+   the commit message.
+   This bear would inspect the changes made in such commits and verify that they
+   can skip CI build that is with respect to modified files and whether the
+   corresponding CI engine allows skipping build or not. An appropriate message
+   is returned if skipping build is not supported.
+   - If the commit type from the output of `VCSCommitBear` indicates that the
+     commit skips CI build, inspection is done inside `CISkipInspectBear`
```

Comment:
Also update this, `VCSCommitBear` is no longer providing `COMMIT_TYPE.ci_skip_commit`, all of the matching is done inside of the `CISkipInspectBear`.